### PR TITLE
Allow selecting between parameterized subobjects

### PR DESCRIPTION
**Pull request by philippjfr**
- [x] Implements https://github.com/pyviz/panel/issues/68
- [x] Implements https://github.com/pyviz/panel/issues/69 (``toggleable_subobjects`` and ``expand_by_default``)
- [x] Implements https://github.com/pyviz/panel/issues/70
- [x] Adds unit tests

### CODE REVIEW

File: panel/param.py [hunk 1]
```diff
@@ -95,9 +96,9 @@ def __init__(self, object, **params):
             params['name'] = object.name
         super(Param, self).__init__(object, **params)
         self._widgets = self._get_widgets()
-        self._widget_box = WidgetBox(*self._widgets.values(), height=self.height,
+        widgets = [widget for widgets in self._widgets.values() for widget in widgets]
+        self._widget_box = WidgetBox(*widgets, height=self.height,
                                      width=self.width, name=self.name)
-
         kwargs = {'name': self.name}
         if self.subpanel_layout is Tabs:
```
Comment (jbednar) [inline panel/param.py]: Should "subpanel" still be appearing in the source code?  I would think it should say "pane" or "viewable"?

File: panel/param.py [hunk 2]
```diff
@@ -260,6 +293,91 @@ def _get_root(self, doc, comm=None):
 
 
 
+class ParamMethod(PaneBase):
+    """
+    ParamMethod panes wrap methods annotated with the param.depends
+    decorator and rerenders the plot when any of the methods parameters
```
Comment (jbednar) [inline panel/param.py]: It's not required for them to be annotated, right?  (Given that the default behavior is to watch all parameters.)

Comment (jbednar) [reply panel/param.py]: "methods" -> "method's"

File: panel/param.py (hunk 1 above)
Comment (philippjfr) [reply panel/param.py]: Neither seem very descriptive, it's specifically about the layout for the parameterized subobjects. I guess `subobject_layout` would work.

Comment (jbednar) [reply panel/param.py]: That's better for this case, thanks.

**Comment by philippjfr** [comment]
I think there's some error on unwatch I haven't yet handled.

**Comment by jbednar** [comment]
Both new options seem to work well.  I did have one problem, which is that I have a parameter that is defined purely to get a subobject pane (with only one value; `constant=True`), and I can't make the parameter not be displayed.  I can set its precedence to make it go away, but then the corresponding subobject param pane also goes away.  Maybe that can be a special case -- If `expand_by_default=True` and there's a constant parameter that would be expanded, don't show it in the main list of parameters (as it then has no reason to be shown, given that its values are now available on screen anyway and it can't be edited or expanded)?  It's an annoying thing to have to do and tricky, but I don't see how else to make it work well in this case.

**Comment by jbednar** [comment]
> It's an annoying thing to have to do and tricky, but I don't see how else to make it work well in this case.

I don't think we need to do this -- I was wanting to see the parameters of the constant subobject to get around https://github.com/ioam/param/issues/287 , but instead I can use a readonly subobject and thereby put a pane for editing its widgets wherever I like.  So the current behavior for hidden parameters (also hiding their subobject parameters) is fine.

**Pull request merged**